sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/timewindow/recurrence/RecurrenceEvaluator.java

```diff
@@ -0,0 +1,158 @@
+        final ZonedDateTime firstDayOfFirstWeek = settings.getStart().minusDays(
+            TimeWindowUtils.daysPassedWeekStart(settings.getStart().getDayOfWeek(), pattern.getFirstDayOfWeek()));
+
+        final long numberOfInterval = Duration.between(firstDayOfFirstWeek, now).toSeconds()
+            / Duration.ofDays((long) interval * RecurrenceConstants.DAYS_PER_WEEK).toSeconds();
+        final ZonedDateTime firstDayOfMostRecentOccurrence = firstDayOfFirstWeek.plusDays(
+            numberOfInterval * (interval * RecurrenceConstants.DAYS_PER_WEEK));
+        final List<DayOfWeek> sortedDaysOfWeek = TimeWindowUtils.sortDaysOfWeek(pattern.getDaysOfWeek(), pattern.getFirstDayOfWeek());
+        final int maxDayOffset = TimeWindowUtils.daysPassedWeekStart(sortedDaysOfWeek.get(sortedDaysOfWeek.size() - 1), pattern.getFirstDayOfWeek());
+        final int minDayOffset = TimeWindowUtils.daysPassedWeekStart(sortedDaysOfWeek.get(0), pattern.getFirstDayOfWeek());
+        ZonedDateTime mostRecentOccurrence = null;
+        int numberOfOccurrences = (int) (numberOfInterval * sortedDaysOfWeek.size()
+            - (sortedDaysOfWeek.indexOf(settings.getStart().getDayOfWeek())));
+
+        // now is after the first week
+        if (now.isAfter(firstDayOfMostRecentOccurrence.plusDays(RecurrenceConstants.DAYS_PER_WEEK))) {
+            numberOfOccurrences += sortedDaysOfWeek.size();
+            mostRecentOccurrence = firstDayOfMostRecentOccurrence.plusDays(maxDayOffset);
+            return new OccurrenceInfo(mostRecentOccurrence, numberOfOccurrences);
+        }
+
+        // day with the min offset in the most recent occurring week
+        ZonedDateTime dayWithMinOffset = firstDayOfMostRecentOccurrence.plusDays(minDayOffset);
+        if (settings.getStart().isAfter(dayWithMinOffset)) {
+            numberOfOccurrences = 0;
+            dayWithMinOffset = settings.getStart();
+        }
+        if (now.isBefore(dayWithMinOffset)) {  // move to last occurrence
+            mostRecentOccurrence = firstDayOfMostRecentOccurrence.minusDays(interval * RecurrenceConstants.DAYS_PER_WEEK).plusDays(maxDayOffset);
+        } else {
+            mostRecentOccurrence = dayWithMinOffset;
+            numberOfOccurrences++;
+
+            for (int i = sortedDaysOfWeek.indexOf(dayWithMinOffset.getDayOfWeek()) + 1; i < sortedDaysOfWeek.size(); i++) {
+                dayWithMinOffset = firstDayOfMostRecentOccurrence.plusDays(
+                    TimeWindowUtils.daysPassedWeekStart(sortedDaysOfWeek.get(i), pattern.getFirstDayOfWeek()));
+                if (now.isBefore(dayWithMinOffset)) {
+                    break;
+                }
+                mostRecentOccurrence = dayWithMinOffset;
+                numberOfOccurrences++;
+            }
+        }
+        return new OccurrenceInfo(mostRecentOccurrence, numberOfOccurrences);
+    }
+
+    private static class OccurrenceInfo {
+        private final ZonedDateTime previousOccurrence;
```

Comment:
I don't undertand why this is a private but it is accessed through `occurrenceInfo.previousOccurrence` in the method of outer class.